proposals/NativeInt.md

```diff
@@ -0,0 +1,263 @@
+| `nint` | `short` |   | `conv.i2` / `conv.ovf.i2` |
+| `nint` | `ushort` |   | `conv.u2` / `conv.ovf.u2` |
+| `nint` | `int` |   | `conv.i4` / `conv.ovf.i4` |
+| `nint` | `uint` |   | `conv.u4` / `conv.ovf.u4` |
+| `nint` | `long` | Y | `conv.i8` / `conv.ovf.i8` |
+| `nint` | `ulong` |   | `conv.u8` / `conv.ovf.u8` |
+| `nint` | `char` |   | `conv.u2` / `conv.ovf.u2` |
+| `nint` | `float` | Y | `conv.r4` |
+| `nint` | `double` | Y | `conv.r8` |
+| `nuint` | `nint` |   | `conv.i` / `conv.ovf.i` |
+| `nuint` | `sbyte` |   | `conv.i1` / `conv.ovf.i1` |
+| `nuint` | `byte` |   | `conv.u1` / `conv.ovf.u1` |
+| `nuint` | `short` |   | `conv.i2` / `conv.ovf.i2` |
+| `nuint` | `ushort` |   | `conv.u2` / `conv.ovf.u2` |
+| `nuint` | `int` |   | `conv.i4` / `conv.ovf.i4` |
+| `nuint` | `uint` |   | `conv.u4` / `conv.ovf.u4` |
+| `nuint` | `long` |   | `conv.i8` / `conv.ovf.i8` |
+| `nuint` | `ulong` | Y | `conv.u8` / `conv.ovf.u8` |
+| `nuint` | `char` |   | `conv.u2` / `conv.ovf.u2.un` |
+| `nuint` | `float` | Y | `conv.r.un conv.r4` |
+| `nuint` | `double` | Y | `conv.r.un conv.r8` |
+
+Conversions between `decimal` and `nint` or `nuint` are not supported because there are no operators on `decimal` that use `native int`.
+There are operators on `decimal` to convert to and from `long` or `ulong` but those are not optimal on 32-bit platforms.
+
+### Operators
+
+The following operators are supported.
+These operators are considered during overload resolution based on normal rules for implicit conversions of arguments.
+
+In cases where there are two overloads, one for `nint` and one for `nuint`, the native int operand types are marked `native`. In other cases the specific types are provided.
+(The IL for each operator includes the variants for `nint` and `nuint` and the variants for `unchecked` and `checked` contexts if different.)
+
+| Unary | Operand | Result | IL |
+|:---:|:---:|:---:|:---:|
+| `-` | `nint` | `nint` | `neg` |
+| `~` | `native` | `native` | `not` |
+
+| Binary | Left | Right | Result | IL |
+|:---:|:---:|:---:|:---:|:---:|
+| `+` | `native` | `native` | `native` | `add` / `add.ovf` / `add.ovf.un` |
+| `-` | `native` | `native` | `native` | `sub` / `sub.ovf` / `sub.ovf.un` |
+| `*` | `native` | `native` | `native` | `mul` / `mul.ovf` / `mul.ovf.un` |
+| `/` | `native` | `native` | `native` | `div` / `div.un` |
+| `%` | `native` | `native` | `native` | `rem` / `rem.un` |
+| `==` | `native` | `native` | `native` | `beq` / `ceq` |
+| `!=` | `native` | `native` | `native` | `bne` |
+| `<` | `native` | `native` | `native` | `blt` / `clt` / `blt.un` / `clt.un` |
+| `<=` | `native` | `native` | `native` | `ble` / `ble.un` |
+| `>` | `native` | `native` | `native` | `bgt` / `cgt` / `bgt.un` / `cgt.un` |
+| `>=` | `native` | `native` | `native` | `bge` / `bge.un` |
+| `&` | `native` | `native` | `native` | `and` |
+| `\|` | `native` | `native` | `native` | `or` |
+| `<<` | `native` | `int` | `native` | `shl` |
```

Comment:
>`<<` [](start = 2, length = 4)

What are the semantics of the shift operators?  For comparison, the spec in [https://github.com/dotnet/csharplang/blob/master/spec/expressions.md#shift-operators](https://github.com/dotnet/csharplang/blob/master/spec/expressions.md#shift-operators) has text that begins "For the predefined operators, the number of bits to shift is computed as follows"